src/core/public/application/ui/scoped_history.ts

```diff
@@ -0,0 +1,290 @@
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import {
+  History,
+  Path,
+  LocationState,
+  LocationDescriptorObject,
+  TransitionPromptHook,
+  UnregisterCallback,
+  LocationListener,
+  Location,
+  Href,
+} from 'history';
+
+/**
+ * A wrapper around a `History` instance that is scoped to a particular base path of the history stack. Behaves
+ * similarly to the `basename` option except that this wrapper hides any history stack entries from outside the scope
+ * of this base path.
+ *
+ * This wrapper also allows Core and Plugins to share a single underlying global `History` instance without exposing
+ * the history of other applications.
+ *
+ * @internal
+ */
+export class ScopedHistory<HistoryLocationState = LocationState>
+  implements History<HistoryLocationState> {
+  /**
+   * Tracks whether or not the user has left this history's scope. All methods throw errors if called after scope has
+   * been left.
+   */
+  private isActive = true;
+  /**
+   * All active listeners on this history instance.
+   */
+  private listeners = new Set<LocationListener<HistoryLocationState>>();
+  /**
+   * Array of the local history stack. Only stores {@link Location.key} to use tracking an index of the current
+   * position of the window in the history stack.
+   */
+  private locationKeys: Array<string | undefined> = [];
+  /**
+   * The key of the current position of the window in the history stack.
+   */
+  private currentLocationKeyIndex: number = 0;
+
+  constructor(
+    private readonly parentHistory: History<HistoryLocationState>,
+    private readonly basePath: string
+  ) {
+    if (!this.parentHistory.location.pathname.startsWith(basePath)) {
+      throw new Error(
+        `Browser location [${this.basePath}] is not currently at expected basePath [${basePath}]`
+      );
+    }
+
+    this.locationKeys.push(this.parentHistory.location.key);
+    this.setupHistoryListener();
+  }
+
+  /**
+   * The number of entries in the history stack, including all entries forwards and backwards from the current location.
+   */
+  public get length() {
+    this.verifyActive();
+    return this.locationKeys.length;
+  }
+
+  /**
+   * The current location of the history stack.
+   */
+  public get location() {
+    this.verifyActive();
+    return this.stripBasePath(this.parentHistory.location);
+  }
+
+  /**
+   * The last action dispatched on the history stack.
+   */
+  public get action() {
+    this.verifyActive();
+    return this.parentHistory.action;
+  }
```

Comment:
Shouldn't we filter parent actions and only return the first action that impacted the scoped history path? Not really sure of the usages / implications here.

Reply:
I'm unsure about how this is used as well, so I'm unsure of the desired behavior here. If the user hits the back button to go from one app to another, should `history.action` be POP or PUSH? Right now it returns POP, which is accurate, but maybe unexpected if the `history.length === 1`?

@Dosant do you have any knowledge about how `history.action` is used, either in our code or React Router?

Reply:
Didn't use action in state sync utils so far. 
Don't know how it is used in react-router

Reply:
Ok I'm going to leave this as-is unless someone has a reason to change it.